gem/gem.py

```diff
@@ -566,6 +582,55 @@ def __new__(cls, i, j):
         self.free_indices = tuple(unique(free_indices))
         return self
 
+    
+class IndexIterator(object):
+    """An iterator whose value is a multi-index (tuple) iterating over the
+    extent of the supplied :class:`.Index` objects in a last index varies
+    fastest (ie 'c') ordering.
+
+    :arg *indices: the indices over whose extent to iterate."""
+    def __init__(self, *indices):
+
+        self.affine_groups = set()
+        for i in indices:
+            if isinstance(i, AffineIndex):
+                try:
+                    pos = tuple(indices.index(g) for g in i.group)
+                except ValueError:
+                    raise ValueError("Only able to iterate over all indices in an affine group at once")
+                self.affine_groups.add((i.group, pos))
+        
+        self.ndindex = numpy.ndindex(tuple(i.extent for i in indices))
+
+    def _affine_groups_legal(self, multiindex):
+        for group, pos in self.affine_groups:
+            if sum(multiindex[p] for p in pos) >= group[0].extent:
+                return False
+        return True
+        
+    def __iter__(self):
+        # Fix this for affine index groups.
+        while True:
+            multiindex = self.ndindex.next()
+            if self._affine_groups_legal(multiindex):
```

Comment:
This isn't the best way to do it, but since it doesn't go to master yet, we can fix it up later.